File: src/main/kotlin/dev/jefy/connectpro/shared/infrastructure/converter/AvailabilityListConverter.java
package dev.jefy.connectpro.shared.infrastructure.converter;


import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import dev.jefy.connectpro.portfolio.domain.vo.Availability;
import jakarta.persistence.AttributeConverter;
import jakarta.persistence.Converter;

/**
 * @author Jôph Yamba
 */
@Converter
public class AvailabilityListConverter implements AttributeConverter<List<Availability>, String> {

    @Override
    public String convertToDatabaseColumn(List<Availability> availabilities) {
        if (availabilities == null || availabilities.isEmpty()) return "";
        return availabilities.stream()
                .map(Enum::name)
                .collect(Collectors.joining(","));
    }

    @Override
    public List<Availability> convertToEntityAttribute(String value) {
        if (value == null || value.isBlank()) return new ArrayList<>();
        return Arrays.stream(value.split(","))
                .map(Availability::valueOf)
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/dev/jefy/connectpro/shared/infrastructure/converter/AvailabilityListConverter.kt
package dev.jefy.connectpro.shared.infrastructure.converter

import dev.jefy.connectpro.portfolio.domain.vo.Availability
import jakarta.persistence.AttributeConverter
import jakarta.persistence.Converter

@Converter
class AvailabilityListConverter : AttributeConverter<MutableList<Availability>, String> {

    override fun convertToDatabaseColumn(availabilities: MutableList<Availability>?): String =
        availabilities
            ?.takeIf { it.isNotEmpty() }
            ?.joinToString(",") { it.name }
            ?: ""

    override fun convertToEntityAttribute(value: String?): MutableList<Availability> =
        value
            ?.takeIf { it.isNotBlank() }
            ?.split(",")
            ?.mapNotNull { img -> img.trim().takeIf { it.isNotEmpty() } }
            ?.map { it ->
                runCatching { Availability.valueOf(it) }
                    .getOrElse { throw IllegalArgumentException("Invalid availability value: $it") }
            }
            ?.toMutableList()
            ?: mutableListOf()
}